universal-login-sdk/lib/core/observers/BalanceObserver.ts

```diff
@@ -1,49 +1,49 @@
-import {providers} from 'ethers';
-import {SupportedToken, ensure, RequiredBalanceChecker, BalanceChecker} from '@universal-login/commons';
-import {ConcurrentDeployment} from '../utils/errors';
 import ObserverRunner from './ObserverRunner';
-
-export type ReadyToDeployCallback = (tokenAddress: string, contractAddress: string) => void;
+import {BalanceChecker, TokenDetails, TokenDetailsWithBalance} from '@universal-login/commons';
+import deepEqual from 'deep-equal';
+import clonedeep from 'lodash.clonedeep';
 
 export class BalanceObserver extends ObserverRunner {
-  private contractAddress?: string;
-  private callback?: ReadyToDeployCallback;
-  private requiredBalanceChecker: RequiredBalanceChecker;
-  private balanceChecker: BalanceChecker;
+  private lastTokenBalances: TokenDetailsWithBalance[] = [];
+  private callbacks: Function[] = [];
 
-  constructor(private supportedTokens: SupportedToken[], provider: providers.Provider) {
+  constructor(private balanceChecker: BalanceChecker, private walletAddress: string, private tokenDetails: TokenDetails[], step: number = 500) {
     super();
-    this.balanceChecker = new BalanceChecker(provider);
-    this.requiredBalanceChecker = new RequiredBalanceChecker(this.balanceChecker);
+    this.step = step;
   }
 
-  async startAndSubscribe(contractAddress: string, callback: ReadyToDeployCallback) {
-    ensure(!this.isRunning(), ConcurrentDeployment);
-    this.contractAddress = contractAddress;
-    this.callback = callback;
-    this.start();
-    return () => {
-      this.contractAddress = undefined;
-      this.stop();
-    };
+  async tick() {
+    await this.checkBalanceNow();
   }
 
-  tick() {
-    return this.checkBalance();
+  async getBalances(): Promise<TokenDetailsWithBalance[]> {
+    return Promise.all(this.tokenDetails.map(async (token: TokenDetails) => {
```

Comment:
Ok, this is actually a bad idea. Promise all is very insidious. 
It can lead to depletion of resources and we can hit rate limiting on popular ethereum node providers. 
We should probably stick to good old for if map can't do it.